packages/ui-components/src/JpThemeProvider.tsx

```diff
@@ -0,0 +1,72 @@
+/*
+ * Copyright 2018-2021 Elyra Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import { IThemeManager } from '@jupyterlab/apputils';
+
+import { createMuiTheme, ThemeProvider } from '@material-ui/core';
+
+import React from 'react';
+
+export interface IJpThemeProviderProps {
+  children: React.ReactElement;
+  themeManager: IThemeManager;
+  updateWidget?: () => void;
+}
+
+export const JpThemeProvider = (
+  props: IJpThemeProviderProps
+): React.ReactElement => {
+  let isLight =
+    props.themeManager.theme &&
+    props.themeManager.isLight(props.themeManager.theme);
+
+  const updateTheme = (): void => {
+    const isLightUpdate =
+      props.themeManager.theme &&
+      props.themeManager.isLight(props.themeManager.theme);
+
+    if (isLightUpdate != isLight) {
+      isLight = isLightUpdate;
+      if (props.updateWidget) {
+        props.updateWidget();
+      }
+    }
+  };
+
+  updateTheme();
+
+  const lightTheme = createMuiTheme({
+    palette: {
+      type: 'light'
+    }
+  });
+
+  const darkTheme = createMuiTheme({
+    palette: {
+      type: 'dark'
+    }
+  });
+
+  if (props.themeManager) {
+    props.themeManager.themeChanged.connect(updateTheme);
```

Comment:
This should go into an effect hook

Reply:
like this? 
```
  useEffect(() => {
    if (props.themeManager) {
      props.themeManager.themeChanged.connect(updateTheme);
    }
  });
```
I just googled it and it seems to the the function equivalent of componentDidMount, why should this be in an effect hook?

Reply:
you also need a dependency array otherwise this will run every render

Reply:
```
useEffect(() => {
    if (themeManager) {
     themeManager.themeChanged.connect(updateTheme);
    }
  }, [themeManager]);
```

Reply:
this will make it so it only connects whenever themeManager or updateTheme change, does this have a disconnect function? because we should do that here as well

Reply:
I'll write up a quick demo with notes on how this works with hooks

Reply:
```js
// The themes should be created outside of the component, otherwise they will be re-created
// every render which could be expensive.
const lightTheme = createMuiTheme({
  palette: {
    type: 'light'
  }
});

const darkTheme = createMuiTheme({
  palette: {
    type: 'dark'
  }
});

function isLightTheme(themeManager) {
  return themeManager.theme && themeManager.isLight(themeManager.theme);
}

function JpThemeProvider({ themeManager, children, updateWidget }) {
  // variables in a component should not be directly modified by side effects. If we want to 
  // modify a variable from an effect/callbacks it should be a state hook.
  // Note: It is normally frowned upon to derive state from props, because the can become 
  // stale. In this case it is okay because we update this value through an effect.
  const [isLight, setIsLight] = useState(isLightTheme(themeManager))

  useEffect(() => {
    // Any type of listeners are side effects and should be set up and torn down inside
    // an effect hook. Otherwise, you will register a listener every render and end up
    // a big leak

    // `updateTheme` is small and isn't called anywhere else so we can define it inside
    // of the effect so we don't need to add it to the dependency array.
    // It's important to note that if this is defined inside of the component and listed in
    // the dependency array the effect would be rerun on every render because a new 
    // function is created every render. You can avoid this by using a `callback` hook
    // which will memoize the function when possible 
    function updateTheme() {
      setIsLight(isLightTheme(themeManager))
      updateWidget()
    }

    if (themeManager) {
      themeManager.themeChanged.connect(updateTheme);
    }
    return () => {
      // disconnect when the when the component is unmounted/the dependency array has 
      // changed and a new connection has been made.
      themeManager.themeChanged.disconnect(updateTheme);
    }
  // This is the dependency array, anytime one of these change this effect will be ran. It will
  // also run when the component is first mounted.
  // Notice, that `updateWidget` is listed here, but not `isLightTheme` or `setIsLight`. This
  // is because any `setState` functions are guaranteed not to change as well as the function
  // defined outside of this component. However, `updateWidget` is a prop and could change 
  // from render to render. (There is an eslint rule that will automatically generate this array for you)
  }, [themeManager, updateWidget])

  return (
    <ThemeProvider theme={isLight ? lightTheme : darkTheme}>
      {children}
    </ThemeProvider>
  )
}
```

Reply:
Here is an example of `updateTheme` with the callback hook:
```
const updateTheme = useCallback(() => {
  setIsLight(isLightTheme(themeManager))
  updateWidget()
}, [themeManager, updateWidget])
```

Reply:
there's also the option to memoize the theme and use it inside of the component:
```
const theme = useMemo(() => {
  return createMuiTheme({
    palette: {
      type: isLight ? 'light' : 'dark'
    }
  });
}, [isLight])
```